src/java/main/org/apache/zookeeper/server/auth/KeyAuthenticationProvider.java

```diff
@@ -0,0 +1,146 @@
+ * limitations under the License.
+ */
+
+package org.apache.zookeeper.server.auth;
+
+import java.io.UnsupportedEncodingException;
+
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.KeeperException.NoNodeException;
+import org.apache.zookeeper.data.Id;
+import org.apache.zookeeper.server.ServerCnxn;
+import org.apache.zookeeper.server.ZooKeeperServer;
+import org.apache.zookeeper.server.ZKDatabase;
+import org.apache.zookeeper.server.DataNode;
+import org.apache.zookeeper.data.Stat;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/*
+ * This class is a sample implementation of being passed the ZooKeeperServer
+ * handle in the constructor, and reading data from zknodes to authenticate.
+ * At a minimum, a real Auth provider would need to override validate() to
+ * e.g. perform certificate validation of auth based a public key.
+ *
+ * See the "Pluggable ZooKeeper authentication" section of the 
+ * "Zookeeper Programmer's Guide" for general details of implementing an
+ * authentication plugin. e.g.
+ * http://zookeeper.apache.org/doc/trunk/zookeeperProgrammers.html#sc_ZooKeeperPluggableAuthentication
+ *
+ * This class looks for a numeric "key" under the /key node.
+ * Authorizaton is granted if the user passes in as authorization a number
+ * which is a multiple of the key value, i.e. 
+ *   (auth % key) == 0
+ * In a real implementation, you might do something like storing a public
+ * key in /key, and using it to verify that auth tokens passed in were signed
+ * by the corresponding private key.
+ *
+ * When the node /key does not exist, any auth token is accepted, so that 
+ * bootstrapping may occur.
+ *
+ */
+public class KeyAuthenticationProvider implements ServerAuthenticationProvider {
+    private static final Logger LOG = LoggerFactory.getLogger(KeyAuthenticationProvider.class);
+    private volatile ZooKeeperServer zks;
+
+    @Override
+    public void setZooKeeperServer(ZooKeeperServer zks) {
+        this.zks = zks;
+        LOG.debug("KeyAuthenticationProvider plugin loaded");
+    }
+
+    public String getScheme() {
+        return "key";
+    }
+
+    public byte[] getKey() {
+        ZKDatabase db = zks.getZKDatabase();
+        if (db != null) {
+            try {
+                Stat stat = new Stat();
+                return db.getData("/key", stat, null);
+            } catch (NoNodeException e) {
+                // ignore
+            }
+        }
+        return null;
+    }
+
+    public boolean validate(byte[] key, byte[] auth) {
+        // perform arbitrary function (auth is a multiple of key)
+        try {
+            String keyStr = new String(key, "UTF-8");
+            String authStr = new String(auth, "UTF-8");
+            int keyVal = Integer.parseInt(keyStr);
+            int authVal = Integer.parseInt(authStr);
+            if (keyVal!=0 && ((authVal % keyVal) != 0)) {
+              return false;
+            }
+        } catch (NumberFormatException nfe) {
```

Comment:
nit: ZK support Java 7, so we could rewrite this line as below (i.e., multi-catch), right?
```
} catch (NumberFormatException|UnsupportedEncodingException ex) {
      return false;
}
```